app/src/util/ingest_utils.py

```diff
@@ -102,3 +105,122 @@ def add_embeddings(
+    assert item_max_seq_length > 0, item_max_seq_length
+    chunks = [
+        intro_sentence + some_list_items
+        for some_list_items in _join_up_to_max_seq_length(
+            list_items, item_max_seq_length, delimiter=join_delimiter
+        )
+    ]
+    assert all(len(tokenize(chunk)) <= token_limit for chunk in chunks)
+    return chunks
+
+
+def _ensure_blank_line_suffix(text: str) -> str:
+    "Return text with a blank line after the text"
+    if not text.endswith("\n\n"):
+        if text.endswith("\n"):
+            text += "\n"
+        else:
+            text += "\n\n"
+    return text
+
+
+def deconstruct_table(
+    markdown: str, has_intro_sentence: bool = False, list_delimiter: str = r"^(\| )"
+) -> tuple[str, str, list[str]]:
+    "Deconstruct a table of items into an introductory sentence, a table header, and a list of items"
+    intro_sentence, list_items = deconstruct_list(markdown, has_intro_sentence, list_delimiter)
+
+    if list_items[1].startswith("| --- |"):
+        table_header = list_items.pop(0) + list_items.pop(0)
+    else:  # Assume no table header
+        table_header = ""
+
+    return intro_sentence, table_header, list_items
+
+
+def reconstruct_table(
+    token_limit: int,
+    intro_sentence: str,
+    table_header: str,
+    table_rows: Sequence[str],
+    join_delimiter: str = "",
+) -> list[str]:
+    "Reconstruct a table of items into chunks (with intro_sentence + table_header) having up to a character limit each."
+
+    # Before the table, there should be a blank line
+    intro_sentence = _ensure_blank_line_suffix(intro_sentence)
+
+    intro = intro_sentence + table_header
+    item_max_seq_length = token_limit - len(tokenize(intro))
+    assert item_max_seq_length > 0, item_max_seq_length
+    print("item_max_seq_length", item_max_seq_length)
+    for row in table_rows:
+        print("row", row)
+    chunks = [
+        intro + some_table_rows
+        for some_table_rows in _join_up_to_max_seq_length(
+            table_rows, item_max_seq_length, delimiter=join_delimiter
+        )
+    ]
+    assert all(len(tokenize(chunk)) <= token_limit for chunk in chunks)
+    return chunks
+
+
+def _join_up_to_max_seq_length(
+    lines: Sequence[str], max_seq_length: int, delimiter: str = " "
+) -> list[str]:
+    chunks = []
+    chunk = ""
+    for line in lines:
+        test_chunk = delimiter.join([chunk, line]) if chunk else line
+        token_count = len(tokenize(test_chunk))
+        logger.debug("%i / %i : %r", token_count, max_seq_length, test_chunk)
+        if token_count > max_seq_length:
+            # Don't use test_chunk; add current chunk; start a new chunk
+            if chunk:
+                chunks.append(chunk)
+
+            if (line_token_count := len(tokenize(line))) <= max_seq_length:
+                chunk = line
+            else:
+                logger.warning("Line is too long (%i tokens): %s", line_token_count, line)
+                chunk = ""
```

Comment:
Am I reading this right, that if an entry in `lines` exceeds the max_seq_length we won't tokenize it at all? (That seems unlikely to occur, but I think the behavior we'd rather have is to have a chunk that exceeds the max_seq_length rather than not ingest it. Maybe not worth implementing handling for that here, but I might upgrade the log message from warning to error so it stands out to us if it does happen)

Reply:
Good catch! This should be an assertion, which fails on https://edd.ca.gov/en/payroll_taxes/Due_Dates_Calendar/ because the list and adjacent table are not being split correctly b/c of interactions with `MarkdownHeaderTextSplitter`. Working on a fix!